File: templating/src/main/java/com/vmware/jsonteng/rules/ForEachRule.java
// Copyright 2019 VMware, Inc.
// SPDX-License-Indentifier: Apache-2.0

package com.vmware.jsonteng.rules;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.vmware.jsonteng.ElementResolver;
import com.vmware.jsonteng.JsonLoader;
import com.vmware.jsonteng.RuleResolver;
import com.vmware.jsonteng.TemplateEngineException;

public class ForEachRule extends RuleBase {
    static final String name = "for-each";

    private final ElementResolver elementResolver;
    private final JsonLoader templateLoader;

    public ForEachRule(RuleResolver ruleResolver) {
        super(ruleResolver);
        this.elementResolver = ruleResolver.getElementResolver();
        this.templateLoader = ruleResolver.getTemplateLoader();
    }

    @Override
    @SuppressWarnings("unchecked")
    public Object process(List<?> ruleTokens, List<Map<String, ?>> bindingDataList) throws TemplateEngineException {
        if (ruleTokens.size() < 2 || ruleTokens.size() > 3) {
            throw new TemplateEngineException(
                    String.format("Rule \"%s\" requires 2 or 3 parameters. Parameters given %s", name, ruleTokens));
        }
        Object dataList = ruleTokens.get(0);
        Object template = ruleTokens.get(1);
        try {
            template = elementResolver.resolve(template, bindingDataList);
        } catch (TemplateEngineException ignore) {
        }
        Object templateJson = templateLoader.load(template.toString());
        List<?> resolvedDataList = (List) elementResolver.resolve(dataList, bindingDataList);
        List<Object> resolvedJson = new ArrayList<>();
        Map<String, Integer> indexData = new HashMap<>();
        for (int i = 0; i < resolvedDataList.size(); i++) {
            Map<String, ?> data = (Map) resolvedDataList.get(i);
            bindingDataList.add(0, data);
            indexData.put("_index_", i);
            bindingDataList.add(0, indexData);
            Object resolvedTemplate = elementResolver.resolve(templateJson, bindingDataList);
            bindingDataList.remove(0);
            bindingDataList.remove(0);
            resolvedJson.add(resolvedTemplate);
        }
        templateLoader.unload(template.toString());
        return resolvedJson;
    }
}

File: templating/core/src/main/python/jsonteng/rules/for_each_rule.py
# Copyright 2019 VMware, Inc.
# SPDX-License-Indentifier: Apache-2.0

from ..exception import TemplateEngineException
from .rule_base import RuleBase


class ForEachRule(RuleBase):
    """
    Apply a list of binding data to a template repeatedly and return the
    resolved templates in a list.
    """
    name = "for-each"

    def __init__(self, rule_resolver):
        """
        Construct this rule.
        :param rule_resolver: Rule resolver
        :type rule_resolver: 'RuleResolver'
        """
        super().__init__(rule_resolver)
        self._element_resolver = rule_resolver.get_element_resolver()
        self._template_loader = rule_resolver.get_template_loader()

    def process(self, rule_tokens, binding_data_list):
        """
        Process this rule.
        :param rule_tokens: Rule arguments.
        :type rule_tokens: 'list'
        :param binding_data_list: Binding data used during the processing.
        :type binding_data_list: 'list'
        :return: JSON object
        :rtype: JSON object
        """
        if len(rule_tokens) < 2 or len(rule_tokens) > 3:
            raise Exception(
                "Rule \"{}\" requires 2 or 3 parameters."
                " Parameters given {}".
                format(ForEachRule.name, rule_tokens))
        data_list = rule_tokens[0]
        try:
            template = self._element_resolver.resolve(
                rule_tokens[1], binding_data_list)
        except TemplateEngineException:
            # If encounter exception, treat the token as the template.
            # The resolve may need a loop dependent binding data.
            template = rule_tokens[1]
        template_json = self._template_loader.load(template)
        resolved_data_list = self._element_resolver.resolve(
            data_list, binding_data_list)
        resolved_json = list()
        for index, data in enumerate(resolved_data_list):
            binding_data_list.insert(0, data)
            binding_data_list.insert(0, {"_index_": index})
            if len(rule_tokens) == 3:
                condition_expr = self._element_resolver.resolve(
                    rule_tokens[2], binding_data_list)
                if not self.safe_eval(condition_expr):
                    binding_data_list.pop(0)
                    binding_data_list.pop(0)
                    continue
            resolved_template = self._element_resolver.resolve(
                template_json, binding_data_list)
            resolved_json.append(resolved_template)
            binding_data_list.pop(0)
            binding_data_list.pop(0)
        self._template_loader.unload(template)
        return resolved_json
